widget/slider.go

```diff
@@ -0,0 +1,289 @@
+package widget
+
+import (
+	"image/color"
+	"math"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/theme"
+)
+
+const (
+	standardScale = 6
+	minLongSide   = 50
+)
+
+var (
+	_ fyne.Draggable = (*Slider)(nil)
+)
+
+// Slider if a widget that can slide between two fixed values.
+type Slider struct {
+	baseWidget
+
+	Value float64
+	Min   float64
+	Max   float64
+	Step  float64
+
+	Vertical  bool
+	OnChanged func()
+}
+
+// Resize sets a new size for a widget.
+func (s *Slider) Resize(size fyne.Size) {
+	s.resize(size, s)
+}
+
+// MinSize returns the smallest size this widget can shrink to
+func (s *Slider) MinSize() fyne.Size {
+	return s.minSize(s)
+}
+
+// Show this widget, if it was previously hidden
+func (s *Slider) Show() {
+	s.show(s)
+}
+
+// Hide this widget, if it was previously visible
+func (s *Slider) Hide() {
+	s.hide(s)
+}
+
+// Move the widget to a new position, relative to its parent.
+func (s *Slider) Move(pos fyne.Position) {
+	s.move(pos, s)
+}
+
+// DragEnd function.
+func (s *Slider) DragEnd() {
+}
+
+// Dragged function.
+func (s *Slider) Dragged(e *fyne.DragEvent) {
+	ok, pos, max := s.wasSliderEvent(&(e.PointEvent))
+
+	if ok {
+		// clamp the position for drags that go out of bounds
+		if pos > max {
+			pos = max
+		} else if pos < 0 {
+			pos = 0
+		}
+
+		s.fireTrigger(pos, max)
+	}
+}
+
+func (s *Slider) wasSliderEvent(e *fyne.PointEvent) (
+	ok bool, pos, max int) {
+
+	render := Renderer(s).(*sliderRenderer)
+
+	tp := render.thumb.Position()
+	ts := render.thumb.Size()
+
+	pad := theme.Padding()
+
+	x := e.Position.X
+	y := e.Position.Y
+
+	if s.Vertical {
+		if x > (tp.X-pad) && x < (tp.X+ts.Width+pad) {
+			// if the cursor was inside the slider area
+			return true, y, render.track.Size().Height
+		}
+	} else {
+		if y > (tp.Y-pad) && y < (tp.Y+ts.Height+pad) {
+			// if the cursor was inside the slider area
+			return true, x, render.track.Size().Width
+		}
+	}
+	return false, 0, 0
+}
+
+func (s *Slider) fireTrigger(pos, max int) {
+	// update value
+	s.updateValue(float64(pos) / float64(max))
+	Refresh(s)
+
+	if s.OnChanged != nil {
+		s.OnChanged()
+	}
+}
+
+func (s *Slider) updateValue(ratio float64) {
+	if s.Vertical {
```

Comment:
The position of the thumb on the track should be proportional to the range between the min and max values. Because the origin is top left of the screen and increases downwards and the min value should be located at the height of the widget ( bottom ) of the track, to increase upwards the inversion needed.
![i](https://user-images.githubusercontent.com/20728117/63103512-e7402c00-bf42-11e9-8151-70e405bde367.png)


That being said I do like the idea of returning the ratio at a place upstream where orientation is already determined. This does require moving some extra logic into it, but make the flow more contiguous.

```go
if s.Vertical {
	if x > (tp.X-pad) && x < (tp.X+ts.Width+pad) {
		if y > t.Height {
			return true, 0.0
		} else if y < 0 {
			return true, 1.0
		} else {
			return true, 1 - float64(y) / float64(t.Height)
		}
	}
} else {
	if y > (tp.Y-pad) && y < (tp.Y+ts.Height+pad) {
		if x > t.Width {
			return true, 1.0
		} else if x < 0 {
			return true, 0.0
		} else {
			return true, float64(x) / float64(t.Width)
		}
	}
}
return false, 0.0
```

Reply:
I agree that the vertical arrangement has reversed logic, but that is a rendering issue and therefore I thought it did not belong in the widget setter logic.